src/shim.c

```diff
@@ -954,6 +966,51 @@ connect_to_proxy(struct cc_shim *shim)
 	return false;
 }
 
+inline void sleep_ms(int ms)
+{
+	struct timespec ts = { 0, ms * 1000000L };
+	nanosleep(&ts, NULL);
+}
+
+/*
+ * Try to re-establish tcp/unix connection with proxy in shim->timeout seconds.
+ *
+ * \param shim \ref cc_shim
+ *
+ * \return true on success, false on failure
+ */
+bool reconnect_to_proxy(struct cc_shim *shim)
+{
+	shim_warning("Reconnecting to cc-proxy (timeout %d s)\n",
+			shim->timeout);
+	int time = 0;
+
+try_again:
```

Comment:
Better to avoid using `goto` when possible for more clarity in the code. Use a loop instead:
```c
do {
   ...
} while ()`
```